rfcs/0135-custom-asserts.md

```diff
@@ -0,0 +1,116 @@
+---
+feature: custom-asserts
+start-date: 2022-09-21
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Allow users to use attribute sets with a boolean attribute `success` and a string attribute `message` instead of a boolean in `assert …; …` expressions.
+
+# Motivation
+[motivation]: #motivation
+
+Since Nix is an untyped language, asserts are often needed to ensure that a function or program works correctly. However, the current assert system is unsuitable for more sophisticated error reporting that aims to inform the user as to what has happened.
+
+Consider a nixpkgs package expression that wants to validate its arguments. Currently, the best way to
+provide a custom error message is to use `assert … -> throw …; …`.
+This method has several disadvantages: Since an implication from a falsehood is always true, you are required
+to invert the condition. Additionally, since the assertion itself is not triggered by the error,
+the function of the `assert` keyword is reduced to providing an imperative shorthand for `seq`. This also means that by default,
+the error location is not printed, and there is no mention of an assert in the error message.
```

Comment:
This is not true. `throw` also retains the location where it is called and it is displayed correctly to the user.

Reply:
hm, I thought it was.